Add BPF and kvstore count/latency metrics

tgraf opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7083)
<!-- Reviewable:end -->

eloycoto commented on pkg/bpf/bpf_linux.go:
```diff
@@ -98,12 +104,15 @@ func UpdateElement(fd int, key, value unsafe.Pointer, flags uint64) error {
 		uintptr(unsafe.Pointer(&uba)),
 		unsafe.Sizeof(uba),
 	)
+	callDuration := time.Since(startTime)
```
I would not use time.Now() and Time.Since() due the monotonic issues that we had seen. I would use spanStat that has a fix for that. 

(We already had problems in the endpoint metrics about this)

tgraf replied:
Good point

eloycoto commented on pkg/bpf/metrics.go:
```diff
@@ -0,0 +1,56 @@
+// Copyright 2019 Authors of Cilium
```
Love this, It has been painful the way that we had. Thanks!

joestringer commented on Documentation/configuration/metrics.rst:
```diff
@@ -67,6 +67,14 @@ Datapath
+
+* ``bpf_syscall_duration_seconds``: Duration of BPF system call performed
+  * Labels: ``operation={lookup, delete, update, objPin, getNextKey, ...}``, ``outcome={success|failure}``
+* ``cilium_bpf_map_ops_total``: Number of BPF map operations performed
```
Any particular reason the second metric above starts with `cilium_`, but the first doesn't? The metrics are coming from cilium anyway, do they need that?

tgraf replied:
Typo

joestringer commented on pkg/bpf/map.go:
```diff
@@ -224,3 +225,21 @@ func GetMapType(t MapType) MapType {
+	regexp.MustCompile(`^(cilium_)(.+)+$`),
+}
+
+func extractCommonName(name string) string {
```
:heart: 

Would be neat to make use of this in the stale map collection handling logic too. But that's an issue for a later PR.

tgraf committed:
bpf: Add map operation metric

Example:
```
cilium_bpf_map_ops_total                              mapName="ct4_global" operation="delete" outcome="success"        12.000000
cilium_bpf_map_ops_total                              mapName="ct6_global" operation="delete" outcome="success"        2.000000
cilium_bpf_map_ops_total                              mapName="ct_any4_global" operation="delete" outcome="success"    169.000000
cilium_bpf_map_ops_total                              mapName="ct_any6_global" operation="delete" outcome="success"    1.000000
cilium_bpf_map_ops_total                              mapName="ep_config" operation="update" outcome="success"         17.000000
cilium_bpf_map_ops_total                              operation="delete" outcome="failure" mapName="ipcache"           1.000000
cilium_bpf_map_ops_total                              mapName="ipcache" operation="update" outcome="success"           42.000000
cilium_bpf_map_ops_total                              mapName="lb4_reverse_nat" operation="update" outcome="success"   5.000000
cilium_bpf_map_ops_total                              mapName="lb4_services" operation="update" outcome="success"      23.000000
cilium_bpf_map_ops_total                              mapName="lxc" operation="delete" outcome="failure"               2.000000
cilium_bpf_map_ops_total                              mapName="lxc" operation="delete" outcome="success"               6.000000
cilium_bpf_map_ops_total                              operation="update" outcome="success" mapName="lxc"               38.000000
cilium_bpf_map_ops_total                              mapName="policy" operation="delete" outcome="failure"            3.000000
cilium_bpf_map_ops_total                              mapName="policy" operation="delete" outcome="success"            1.000000
cilium_bpf_map_ops_total                              operation="update" outcome="success" mapName="policy"            591.000000
```

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
kvstore: Add kvstore operation duration metric

Introduces a new metric kvstore_operations_seconds which records the duration
of all kvstore operations labelled by scope, kind, action and outcome.

Also adds an "outcome" label to differentiate between successful and failed
kvstore operations.

Example:
```
cilium_kvstore_operations_duration_seconds            action="CreateOnly" kind="set" outcome="success" scope="identities/v1"       0.210135
cilium_kvstore_operations_duration_seconds            action="DeletePrefix" kind="delete" outcome="success" scope="cilium-net/o"   0.002148
cilium_kvstore_operations_duration_seconds            outcome="success" scope="identities/v1" action="ListPrefix" kind="read"      0.010153
cilium_kvstore_operations_duration_seconds            action="Update" kind="set" outcome="success" scope="identities/v1"           0.313802
cilium_kvstore_operations_duration_seconds            action="Update" kind="set" outcome="success" scope="ip/v1"                   0.739509
cilium_kvstore_operations_duration_seconds            action="Update" kind="set" outcome="success" scope="nodes/v1"                0.037060
```

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Add syscall duration metric

```
cilium_bpf_syscall_duration_seconds                   outcome="success" operation="create"                           0.039385
cilium_bpf_syscall_duration_seconds                   operation="delete" outcome="failure"                           0.000001
cilium_bpf_syscall_duration_seconds                   outcome="success" operation="delete"                           0.000556
cilium_bpf_syscall_duration_seconds                   operation="getNextKey" outcome="failure"                       0.011878
cilium_bpf_syscall_duration_seconds                   operation="getNextKey" outcome="success"                       0.040337
cilium_bpf_syscall_duration_seconds                   outcome="failure" operation="lookup"                           0.000004
cilium_bpf_syscall_duration_seconds                   operation="lookup" outcome="success"                           0.017185
cilium_bpf_syscall_duration_seconds                   operation="objGet" outcome="success"                           0.001013
cilium_bpf_syscall_duration_seconds                   operation="objPin" outcome="success"                           0.000164
cilium_bpf_syscall_duration_seconds                   operation="update" outcome="success"                           0.000914
```

Signed-off-by: Thomas Graf <thomas@cilium.io>

joestringer reviewed (approved):
One minor nit.

tgraf committed:
spanstat: Simplify usage

Allow to use spanstat this way:
```
duration := spanstat.Start()
err := work()
duration.EndError(err)
```

Signed-off-by: Thomas Graf <thomas@cilium.io>
